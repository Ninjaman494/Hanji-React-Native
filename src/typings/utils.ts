export type FormikContext = Record<
  string,
  string | number | boolean | Record<string, unknown>
>;
