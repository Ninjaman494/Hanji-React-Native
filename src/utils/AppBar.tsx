import React, { useState } from "react";
import { Appbar, IconButton, TextInput } from "react-native-paper";
import { StyleSheet, View } from "react-native";
import { useHistory } from "react-router";

const theme = {
  colors: {
    text: "white",
    color: "white",
    primary: "white",
    background: "transparent",
  },
};

const style = StyleSheet.create({
  input: {
    borderRadius: 0,
    borderTopLeftRadius: 0,
    borderTopRightRadius: 0,
    height: 57,
    overflow: "hidden",
  },
});

const AppBar: React.FC = () => {
  const [showSearch, toggleSearch] = useState(false);
  const [searchQuery, setSearchQuery] = useState<string>("");
  const history = useHistory();

  const doSearch = () => {
    if (searchQuery) {
      history.push(`/search?query=${searchQuery}`);
    }
  };

  return (
    <Appbar.Header>
      <Appbar.Content title="Display Page" />
      {showSearch && (
        <View style={{ flexDirection: "row" }}>
          <TextInput
            value={searchQuery}
            onChangeText={setSearchQuery}
            onSubmitEditing={doSearch}
            placeholder="Search in Korean or English..."
            theme={theme}
            style={style.input}
            underlineColor="transparent"
          />
          <IconButton
            icon="window-close"
            color="white"
            onPress={() => toggleSearch(!showSearch)}
          />
        </View>
      )}
      {!showSearch && (
        <Appbar.Action
          icon="magnify"
          onPress={() => toggleSearch(!showSearch)}
        />
      )}
      <Appbar.Action icon="dots-vertical" onPress={() => {}} />
    </Appbar.Header>
  );
};

export default AppBar;
