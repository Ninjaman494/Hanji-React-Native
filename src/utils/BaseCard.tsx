import React from "react";
import { Card, useTheme } from "react-native-paper";
import { StyleProp, ViewStyle } from "react-native";

export interface BaseCardProps {
  title?: string;
  style?: StyleProp<ViewStyle>;
}

const BaseCard: React.FC<BaseCardProps> = ({ title, children, style }) => {
  const { colors } = useTheme();

  return (
    <Card elevation={5} style={style}>
      {title && (
        <Card.Title
          title={title}
          titleStyle={{ color: colors.primary, fontSize: 16 }}
        />
      )}
      <Card.Content style={{ paddingHorizontal: 0 }}>{children}</Card.Content>
    </Card>
  );
};

export default BaseCard;
