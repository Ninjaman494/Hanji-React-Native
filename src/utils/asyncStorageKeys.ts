/**
 * AsyncStorage key for user id which is synced
 * between Sentry,RevenueCat, and Firebase
 */
export const USER_ID_KEY = "USER_ID";
/**
 * AsyncStorage key for favorite conjugations,
 * which is stored as a JSON string
 */
export const FAVORITES_KEY = "@favorites_Key";

/**
 * AsyncStorage key for the previous version of
 * the app. Used for showing changelog
 */
export const LAST_VERSION_KEY = "LAST_VERSION";
/**
 * AsyncStorage key for the number of times the
 * user has opened the app.
 */
export const SESSIONS_KEY = "NUM_SESSIONS";

/**
 * AsyncStorage key for whether we've fetched
 * the user's purchase history from RevenueCat
 */
export const RESTORED_KEY = "RESTORED_PURCHASES";

/**
 * AsyncStorage key for whether the rating request
 * dialog has been shown
 */
export const SHOWN_KEY = "ALREADY_SHOWN";
