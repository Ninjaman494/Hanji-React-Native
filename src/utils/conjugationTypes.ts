export enum Tense {
  PRESENT,
  PAST,
  FUTURE,
  NONE,
}

export enum Formality {
  INFORMAL_LOW = "informal low",
  INFORMAL_HIGH = "informal high",
  FORMAL_LOW = "formal low",
  FORMAL_HIGH = "formal high",
  NONE = "none",
}

export enum ConjugationType {
  CONNECTIVE_IF = "connective if",
  CONNECTIVE_AND = "connective and",
  CONNECTIVE_BUT = "connective but",
  DECLARATIVE_PRESENT = "declarative present",
  DECLARATIVE_PAST = "declarative past",
  DECLARATIVE_FUTURE = "declarative future",
  DETERMINER_PRESENT = "determiner present",
  DETERMINER_PAST = "determiner past",
  DETERMINER_FUTURE = "determiner future",
  IMPERATIVE = "imperative",
  INTERROGATIVE_PRESENT = "interrogative present",
  INTERROGATIVE_PAST = "interrogative past",
  NOMINAL_ING = "nominal ing",
  PROPOSITIVE = "propositive",
  SUPPOSITIVE = "suppositive",
}

type FormalityMinusNone = Exclude<Formality, Formality.NONE>;

export type ConjugationName =
  | ConjugationType.CONNECTIVE_IF
  | ConjugationType.CONNECTIVE_AND
  | ConjugationType.CONNECTIVE_BUT
  | ConjugationType.NOMINAL_ING
  | ConjugationType.DETERMINER_PRESENT
  | ConjugationType.DETERMINER_PAST
  | ConjugationType.DETERMINER_FUTURE
  | `${ConjugationType.DECLARATIVE_PRESENT} ${FormalityMinusNone}`
  | `${ConjugationType.DECLARATIVE_PAST} ${FormalityMinusNone}`
  | `${ConjugationType.DECLARATIVE_FUTURE} ${FormalityMinusNone}`
  | `${ConjugationType.IMPERATIVE} ${FormalityMinusNone}`
  | `${ConjugationType.INTERROGATIVE_PRESENT} ${FormalityMinusNone}`
  | `${ConjugationType.INTERROGATIVE_PAST} ${FormalityMinusNone}`
  | `${ConjugationType.PROPOSITIVE} ${FormalityMinusNone}`
  | `${ConjugationType.SUPPOSITIVE} ${FormalityMinusNone}`;
