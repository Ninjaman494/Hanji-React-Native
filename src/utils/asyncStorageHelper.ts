import AsyncStorage from "@react-native-async-storage/async-storage";
import { PageName } from "typings/navigation";

/**
 * AsyncStorage key for user id which is synced
 * between Sentry,RevenueCat, and Firebase
 */
export const USER_ID_KEY = "USER_ID";
/**
 * AsyncStorage key for favorite conjugations,
 * which is stored as a JSON string
 */
export const FAVORITES_KEY = "@favorites_Key";

/**
 * AsyncStorage key for the previous version of
 * the app. Used for showing changelog
 */
export const LAST_VERSION_KEY = "LAST_VERSION";
/**
 * AsyncStorage key for the number of times the
 * user has opened the app.
 */
export const SESSIONS_KEY = "NUM_SESSIONS";

/**
 * AsyncStorage key for whether we've fetched
 * the user's purchase history from RevenueCat
 */
export const RESTORED_KEY = "RESTORED_PURCHASES";

/**
 * AsyncStorage key for whether the rating request
 * dialog has been shown
 */
export const SHOWN_KEY = "ALREADY_SHOWN";

/**
 * Has the honorific toggle on Conjugations Page
 * been used
 */
export const HONORIFIC_TOGGLED_KEY = "HONORIFIC_TOGGLED";

export type PageVisitKey = `${PageName}_VISIT_COUNT`;

export type PopupShownKey = `${string}_POPUP_SHOWN`;

export type AsyncStorageKey =
  | typeof USER_ID_KEY
  | typeof FAVORITES_KEY
  | typeof LAST_VERSION_KEY
  | typeof SESSIONS_KEY
  | typeof RESTORED_KEY
  | typeof SHOWN_KEY
  | typeof HONORIFIC_TOGGLED_KEY
  | PageVisitKey
  | PopupShownKey;

export const convertToInt = (val: string | null) => (val ? parseInt(val) : 0);

export const convertToObj = (val: string | null) =>
  val ? JSON.parse(val) : null;

export const convertToBool = (val: string | null) => val === "true";

export const getAsyncStorage = async <
  T extends "string" | "number" | "boolean" | "object"
>(
  key: AsyncStorageKey,
  type: T
): Promise<
  T extends "string"
    ? string | null
    : T extends "number"
    ? number
    : T extends "boolean"
    ? boolean
    : unknown
> => {
  const val = await AsyncStorage.getItem(key);
  let newVal;
  switch (type) {
    case "string":
      newVal = val as string;
      break;
    case "number":
      newVal = val ? parseInt(val) : 0;
      break;
    case "boolean":
      newVal = val === "true";
      break;
    default:
      newVal = val ? JSON.parse(val) : null;
      break;
  }
  return newVal;
};

export const setAsyncStorage = async (
  key: AsyncStorageKey,
  val: string | number | boolean | object
) => {
  const valType = typeof val;

  let newVal: string;
  switch (valType) {
    case "string":
      newVal = val as string;
      break;
    case "number":
      newVal = val.toString();
      break;
    case "boolean":
      newVal = val ? "true" : "false";
      break;
    default:
      newVal = JSON.stringify(val);
      break;
  }

  await AsyncStorage.setItem(key, newVal);
};
