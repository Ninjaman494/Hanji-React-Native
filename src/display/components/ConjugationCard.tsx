import React from "react";
import { StyleSheet } from "react-native";
import { Text, useTheme } from "react-native-paper";
import { Col, Row, Grid } from "react-native-easy-grid";
import { Conjugation } from "../../hooks/useConjugations";
import BaseCard, { BaseCardProps } from "../../components/BaseCard";

export interface ConjugationCardProps extends BaseCardProps {
  conjugations: Conjugation[];
}

const ConjugationCard: React.FC<ConjugationCardProps> = ({
  conjugations,
  ...rest
}) => {
  const { padding, textSizes } = useTheme();

  const style = StyleSheet.create({
    conjView: {
      paddingHorizontal: padding.horizontal,
    },
    rowView: {
      marginBottom: padding.vertical,
    },
    text: {
      fontSize: textSizes.regular,
    },
    divider: {
      fontSize: textSizes.regular,
      textAlign: "center",
    },
  });

  return (
    <BaseCard {...rest}>
      <Grid style={style.conjView}>
        {conjugations.map((conjugation) => (
          <Row style={style.rowView}>
            <Col>
              <Text style={style.text}>{conjugation.name}</Text>
            </Col>
            <Col>
              <Text style={style.divider}>:</Text>
            </Col>
            <Col>
              <Text style={style.text}>{conjugation.conjugation}</Text>
            </Col>
          </Row>
        ))}
      </Grid>
    </BaseCard>
  );
};

export default ConjugationCard;
