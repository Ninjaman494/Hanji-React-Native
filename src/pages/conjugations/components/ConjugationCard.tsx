import { useNavigation } from "@react-navigation/native";
import { BaseCard, BaseCardProps } from "components";
import { Conjugation } from "hooks/useConjugations";
import React from "react";
import { StyleSheet } from "react-native";
import { Col, Grid, Row } from "react-native-easy-grid";
import { Button, Card, Text, useTheme } from "react-native-paper";
import { NavigationProps } from "typings/navigation";
import { Formality } from "utils/conjugationTypes";
import logEvent, { LOG_EVENT } from "utils/logEvent";

export type ConjugationCardProps = BaseCardProps & {
  conjugations: Conjugation[];
  onPress?: () => void;
};

const ConjugationCard: React.FC<ConjugationCardProps> = ({
  conjugations,
  onPress,
  ...rest
}) => {
  const { colors, padding, textSizes } = useTheme();
  const navigation = useNavigation<NavigationProps>();

  const style = StyleSheet.create({
    conjView: {
      paddingHorizontal: padding.horizontal,
    },
    rowView: {
      marginBottom: padding.vertical,
    },
    text: {
      fontSize: textSizes.regular,
    },
    divider: {
      fontSize: textSizes.regular,
      textAlign: "center",
    },
    actions: {
      justifyContent: "flex-end",
      paddingBottom: 0,
    },
  });

  return (
    <BaseCard {...rest}>
      <Grid style={style.conjView}>
        {conjugations.map((conjugation, index) => (
          <Row
            style={style.rowView}
            key={index}
            onPress={async () => {
              await logEvent({
                type: LOG_EVENT.SELECT_CONJUGATION,
                params: {
                  name: conjugation.name,
                  conjugation: conjugation.conjugation,
                  honorific: conjugation.honorific,
                },
              });
              navigation.push("ConjInfo", { conjugation });
            }}
          >
            <Col size={5}>
              <Text style={style.text}>
                {conjugation.speechLevel === Formality.NONE
                  ? conjugation.name
                  : conjugation.speechLevel.replace(/_/g, " ").toLowerCase()}
              </Text>
            </Col>
            <Col>
              <Text style={style.divider}>:</Text>
            </Col>
            <Col size={5}>
              <Text style={style.text}>{conjugation.conjugation}</Text>
            </Col>
          </Row>
        ))}
      </Grid>
      {onPress && (
        <Card.Actions style={style.actions}>
          <Button onPress={onPress} color={colors.accent}>
            See all
          </Button>
        </Card.Actions>
      )}
    </BaseCard>
  );
};

export default ConjugationCard;
