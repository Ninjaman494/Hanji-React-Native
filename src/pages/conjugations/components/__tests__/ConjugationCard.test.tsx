import "react-native";
import React from "react";
import { render } from "@testing-library/react-native";
import ConjugationCard from "../ConjugationCard";
import { ConjugationName, Formality, Tense } from "utils/conjugationTypes";

describe("BaseCard component", () => {
  const baseConjugation = {
    type: "type",
    tense: Tense.PRESENT,
    speechLevel: Formality.INFORMAL_LOW,
    honorific: false,
    pronunciation: "pronunciation",
    romanization: "romanization",
    reasons: [],
  };

  const props = {
    conjugations: [
      {
        name: "name 1" as ConjugationName,
        conjugation: "conj 1",
        ...baseConjugation,
      },
      {
        name: "name 2" as ConjugationName,
        conjugation: "conj 2",
        ...baseConjugation,
      },
      {
        name: "name 3" as ConjugationName,
        conjugation: "conj 3",
        ...baseConjugation,
      },
    ],
  };

  it("displays conjugations", () => {
    const component = render(<ConjugationCard {...props} />);
    expect(component).toMatchSnapshot();
  });
});
