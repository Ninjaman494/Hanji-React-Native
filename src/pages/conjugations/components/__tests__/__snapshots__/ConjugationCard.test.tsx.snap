// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConjugationCard displays conjugations 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={false}
    nativeID="animatedComponent"
    style={
      Object {
        "backgroundColor": "#ffffff",
        "borderColor": "rgba(0, 0, 0, 0.12)",
        "borderRadius": 10,
        "elevation": 5,
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 4,
          "width": 0,
        },
        "shadowOpacity": 0.24,
        "shadowRadius": 5,
      }
    }
  >
    <View
      accessibilityState={
        Object {
          "disabled": true,
        }
      }
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "flexGrow": 1,
          "flexShrink": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "paddingHorizontal": 16,
            },
            undefined,
            Object {
              "paddingBottom": 8,
              "paddingHorizontal": 0,
            },
          ]
        }
      >
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "column",
              "paddingHorizontal": 16,
            }
          }
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            nativeID="animatedComponent"
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "flex": 1,
                "flexDirection": "row",
                "marginBottom": 8,
                "opacity": 1,
              }
            }
          >
            <View
              onPress={[Function]}
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                  "marginBottom": 8,
                }
              }
            >
              <View
                size={5}
                style={
                  Object {
                    "flex": 5,
                    "flexDirection": "column",
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#000000",
                        "fontFamily": "System",
                        "fontWeight": "400",
                      },
                      Object {
                        "textAlign": "left",
                      },
                      Object {
                        "fontSize": 20,
                      },
                    ]
                  }
                >
                  informal low
                </Text>
              </View>
              <View
                style={
                  Object {
                    "flex": 1,
                    "flexDirection": "column",
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#000000",
                        "fontFamily": "System",
                        "fontWeight": "400",
                      },
                      Object {
                        "textAlign": "left",
                      },
                      Object {
                        "fontSize": 20,
                        "textAlign": "center",
                      },
                    ]
                  }
                >
                  :
                </Text>
              </View>
              <View
                size={5}
                style={
                  Object {
                    "flex": 5,
                    "flexDirection": "column",
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#000000",
                        "fontFamily": "System",
                        "fontWeight": "400",
                      },
                      Object {
                        "textAlign": "left",
                      },
                      Object {
                        "fontSize": 20,
                      },
                    ]
                  }
                >
                  conj 1
                </Text>
              </View>
            </View>
          </View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            nativeID="animatedComponent"
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "flex": 1,
                "flexDirection": "row",
                "marginBottom": 8,
                "opacity": 1,
              }
            }
          >
            <View
              onPress={[Function]}
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                  "marginBottom": 8,
                }
              }
            >
              <View
                size={5}
                style={
                  Object {
                    "flex": 5,
                    "flexDirection": "column",
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#000000",
                        "fontFamily": "System",
                        "fontWeight": "400",
                      },
                      Object {
                        "textAlign": "left",
                      },
                      Object {
                        "fontSize": 20,
                      },
                    ]
                  }
                >
                  informal high
                </Text>
              </View>
              <View
                style={
                  Object {
                    "flex": 1,
                    "flexDirection": "column",
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#000000",
                        "fontFamily": "System",
                        "fontWeight": "400",
                      },
                      Object {
                        "textAlign": "left",
                      },
                      Object {
                        "fontSize": 20,
                        "textAlign": "center",
                      },
                    ]
                  }
                >
                  :
                </Text>
              </View>
              <View
                size={5}
                style={
                  Object {
                    "flex": 5,
                    "flexDirection": "column",
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#000000",
                        "fontFamily": "System",
                        "fontWeight": "400",
                      },
                      Object {
                        "textAlign": "left",
                      },
                      Object {
                        "fontSize": 20,
                      },
                    ]
                  }
                >
                  conj 2
                </Text>
              </View>
            </View>
          </View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            nativeID="animatedComponent"
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "flex": 1,
                "flexDirection": "row",
                "marginBottom": 8,
                "opacity": 1,
              }
            }
          >
            <View
              onPress={[Function]}
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                  "marginBottom": 8,
                }
              }
            >
              <View
                size={5}
                style={
                  Object {
                    "flex": 5,
                    "flexDirection": "column",
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#000000",
                        "fontFamily": "System",
                        "fontWeight": "400",
                      },
                      Object {
                        "textAlign": "left",
                      },
                      Object {
                        "fontSize": 20,
                      },
                    ]
                  }
                >
                  formal low
                </Text>
              </View>
              <View
                style={
                  Object {
                    "flex": 1,
                    "flexDirection": "column",
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#000000",
                        "fontFamily": "System",
                        "fontWeight": "400",
                      },
                      Object {
                        "textAlign": "left",
                      },
                      Object {
                        "fontSize": 20,
                        "textAlign": "center",
                      },
                    ]
                  }
                >
                  :
                </Text>
              </View>
              <View
                size={5}
                style={
                  Object {
                    "flex": 5,
                    "flexDirection": "column",
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#000000",
                        "fontFamily": "System",
                        "fontWeight": "400",
                      },
                      Object {
                        "textAlign": "left",
                      },
                      Object {
                        "fontSize": 20,
                      },
                    ]
                  }
                >
                  conj 3
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
